Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2592,52 +2593,51 @@ extension BrowserViewController: HistoryStateHelperDelegate {
 extension BrowserViewController {
 
     func addCustomSearchButtonToWebView(_ webView: WKWebView) {
-        // For now we're going to just not add the custom search button to the web view
-        // TODO: #586 Re-enable custom search engines button or remove entirely
-        return
-        
-        /*
-        //check if the search engine has already been added.
-        let domain = webView.url?.domainURL.host
-        let matches = self.profile.searchEngines.orderedEngines.filter {$0.shortName == domain}
-        if !matches.isEmpty {
-            self.customSearchEngineButton.tintColor = UIColor.Photon.Grey50
-            self.customSearchEngineButton.isUserInteractionEnabled = false
-        } else {
-            self.customSearchEngineButton.tintColor = UIConstants.SystemBlueColor
-            self.customSearchEngineButton.isUserInteractionEnabled = true
-        }
-
-        /*
-         This is how we access hidden views in the WKContentView
-         Using the public headers we can find the keyboard accessoryView which is not usually available.
-         Specific values here are from the WKContentView headers.
-         https://github.com/JaviSoto/iOS9-Runtime-Headers/blob/master/Frameworks/WebKit.framework/WKContentView.h
-        */
-        guard let webContentView = UIView.findSubViewWithFirstResponder(webView) else {
-            /*
-             In some cases the URL bar can trigger the keyboard notification. In that case the webview isnt the first responder
-             and a search button should not be added.
-             */
-            return
-        }
-
-        guard let input = webContentView.perform(#selector(getter: UIResponder.inputAccessoryView)),
-            let inputView = input.takeUnretainedValue() as? UIInputView,
-            let nextButton = inputView.value(forKey: "_nextItem") as? UIBarButtonItem,
-            let nextButtonView = nextButton.value(forKey: "view") as? UIView else {
-                //failed to find the inputView instead lets use the inputAssistant
-                addCustomSearchButtonToInputAssistant(webContentView)
-                return
-            }
-            inputView.addSubview(self.customSearchEngineButton)
-            self.customSearchEngineButton.snp.remakeConstraints { make in
-                make.leading.equalTo(nextButtonView.snp.trailing).offset(20)
-                make.width.equalTo(inputView.snp.height)
-                make.top.equalTo(nextButtonView.snp.top)
-                make.height.equalTo(inputView.snp.height)
+        self.openSearchLinkDict = nil
+        webView.evaluateJavaScript(OpenSearchEngine.fetchOpenSearchLinkScript) { (result, _) in
+            if let dict = (result as? String)?.jsonObject() as? [String: String] {
```

Comment:
I think a `guard let` might be better served here. 

Also you might get away with using a JSON Decoder here since strings automatically conform.
```swift
let decoder = JSONDecoder()
guard let dict = try? decoder.decode([String: String].self, from: result) else { return }
```

Reply:
The problem is `result` is of type `Any` while the expected type is String(the script returns a string). The Decoder requires a `Data` type. Hence the helper function. Since the function is in String Extension I cannot hardcode it to `[String: String]` type.

Reply:
Fair enough.